core/GlobalSubstitution.h

```diff
@@ -50,6 +59,49 @@ class GlobalSubstitution {
     const int toGlobalStateId;
 };
 
+/**
+ * GlobalSubstitution, but lazily populates `nameSubstitution` _and_ builds up a UsageHash for the file.
+ * Used in the hashing package as a part of the AST hashing process, which rewrites ASTs from the main GlobalState into
+ * ASTs for new and empty GlobalStates.
+ */
+class LazyGlobalSubstitution final {
+    const core::GlobalState &fromGS;
+    core::GlobalState &toGS;
+
+    UnorderedMap<core::NameRef, core::NameRef> nameSubstitution;
+    core::UsageHash acc;
+
+    NameRef defineName(NameRef from, bool allowSameFromTo);
+
+public:
+    LazyGlobalSubstitution(const GlobalState &fromGS, GlobalState &toGS);
+    ~LazyGlobalSubstitution() = default;
+
+    NameRef substitute(NameRef from, bool allowSameFromTo = false) {
+        if (&fromGS == &toGS) {
+            return from;
+        }
+
+        auto it = nameSubstitution.find(from);
+        if (it == nameSubstitution.end()) {
+            return defineName(from, allowSameFromTo);
+        }
+        return it->second;
+    }
+
+    NameRef substituteConstant(NameRef from, bool allowSameFromTo = false) {
```

Comment:
These new semantically-enhanced:TM: substitute functions allow us to build up the LSP UsageHash (used for faster Find All References) in the same pass that rewrites the tree. Win/win!

Note that `substituteConstant` is a bit of a misnomer. It's not actually a Constant name. It could be a method name. It's called "substituteConstant" because the UsageHash has a field called `constants`.

Perhaps I should use `substituteSymbolName` instead?